File: sportix-backend/app/services/pulse_service.py
from datetime import datetime, timezone
from appwrite.query import Query as Q
from appwrite.id import ID
from app.core.appwrite import db, DB_ID
from app.core.config import settings
from typing import Optional


async def get_pulse(user_id: str) -> dict:
    res = db.list_documents(
        DB_ID, settings.collection_pulse_scores,
        queries=[Q.equal("user_id", user_id), Q.limit(1)],
    )
    docs = res.get("documents", [])
    if not docs:
        # Fallback check profile pulse_score
        profile_pulse = 100.0
        try:
            p_doc = db.get_document(DB_ID, settings.collection_users, user_id)
            if p_doc and p_doc.get("pulse_score") is not None:
                profile_pulse = float(p_doc.get("pulse_score"))
        except Exception:
            pass
        return _default_pulse(user_id, profile_pulse)

    doc = docs[0]
    total_pulse = doc.get("total_pulse", 100.0)
    level_data = await get_level(user_id)

    return {
        "user_id": user_id,
        "total_pulse": total_pulse,
        "totalPulse": total_pulse,
        "match_performance": doc.get("match_performance", 0),
        "matchPerformance": doc.get("match_performance", 0),
        "consistency": doc.get("consistency", 0),
        "team_chemistry": doc.get("team_chemistry", 0),
        "teamChemistry": doc.get("team_chemistry", 0),
        "reliability": doc.get("reliability", 0),
        "activity": doc.get("activity", 0),
        "leadership": doc.get("leadership", 0),
        "tier": doc.get("tier", "contender"),
        "level": level_data.get("current_level") or level_data.get("level", 1),
        "level_progress_percent": level_data.get("progressPercent", 0),
        "prestige_rank": _get_prestige(total_pulse),
        "updated_at": doc.get("updated_at"),
        "created_at": doc.get("created_at"),
    }


async def get_history(user_id: str, limit: int = 30) -> dict:
    return db.list_documents(
        DB_ID, settings.collection_pulse_history,
        queries=[Q.equal("user_id", user_id), Q.limit(limit), Q.order_desc("created_at")],
    )


async def get_level(user_id: str) -> dict:
    res = db.list_documents(
        DB_ID, settings.collection_user_levels,
        queries=[Q.equal("user_id", user_id), Q.limit(1)],
    )
    docs = res.get("documents", [])
    if not docs:
        return {"current_level": 1, "level": 1, "xp": 0, "progressPercent": 0, "prestigeRank": "Rookie"}
    doc = docs[0]
    level = doc.get("current_level", 1)
    current_pulse = doc.get("current_pulse", 0)
    pulse_for_next = doc.get("pulse_for_next", 200)
    progress = round((current_pulse / max(1, pulse_for_next)) * 100, 1)
    doc["progressPercent"] = min(99.9, max(0, progress))
    doc["level"] = level
    return doc


async def get_level_history(user_id: str) -> dict:
    return db.list_documents(
        DB_ID, settings.collection_level_history,
        queries=[Q.equal("user_id", user_id), Q.limit(50), Q.order_desc("created_at")],
    )


async def get_ssr(user_id: str, sport: Optional[str] = None) -> dict:
    """Sport-Specific Rating — filtered pulse stats for a given sport."""
    pulse = await get_pulse(user_id)
    return {
        "user_id": user_id,
        "sport": sport,
        "overall_ssr": pulse.get("total_pulse", 100),
        "match_performance": pulse.get("match_performance", 0),
        "consistency": pulse.get("consistency", 0),
        "reliability": pulse.get("reliability", 0),
    }


async def award_pulse(user_id: str, source: str, amount: float, reason: Optional[str] = None, reference_id: Optional[str] = None):
    """Add pulse points to a user, update pulse_scores, sync profile, and log to pulse_history."""
    now_iso = datetime.now(timezone.utc).isoformat()
    res = db.list_documents(
        DB_ID, settings.collection_pulse_scores,
        queries=[Q.equal("user_id", user_id), Q.limit(1)],
    )
    new_total = 100.0
    if res.get("documents"):
        doc = res["documents"][0]
        new_total = min(1000.0, float(doc.get("total_pulse", 100.0)) + amount)
        db.update_document(
            DB_ID, settings.collection_pulse_scores, doc["$id"],
            {"total_pulse": new_total, "updated_at": now_iso}
        )
    else:
        new_total = min(1000.0, 100.0 + amount)
        db.create_document(
            DB_ID, settings.collection_pulse_scores, ID.unique(),
            {
                "user_id": user_id,
                "total_pulse": new_total,
                "match_performance": 0,
                "consistency": 0,
                "team_chemistry": 0,
                "reliability": 0,
                "activity": 0,
                "leadership": 0,
                "tier": "pulse_elite" if new_total >= 900 else "elite" if new_total >= 800 else "contender",
                "created_at": now_iso,
                "updated_at": now_iso,
            }
        )

    # Sync profiles collection
    try:
        db.update_document(DB_ID, settings.collection_users, user_id, {"pulse_score": int(new_total)})
    except Exception:
        pass

    # Log to pulse_history
    try:
        db.create_document(DB_ID, settings.collection_pulse_history, ID.unique(), {
            "user_id": user_id,
            "source": source,
            "delta": amount,
            "reason": reason or "",
            "reference_id": reference_id or "",
            "score_after": new_total,
            "created_at": now_iso,
            "updated_at": now_iso,
        })
    except Exception as e:
        print(f"[!] pulse_history log warning: {e}")

    # Check for level up
    await _check_level_up(user_id, new_total)


async def _check_level_up(user_id: str, total_pulse: float):
    """Auto-level-up based on total pulse thresholds."""
    level = _pulse_to_level(total_pulse)
    now_iso = datetime.now(timezone.utc).isoformat()
    try:
        res = db.list_documents(
            DB_ID, settings.collection_user_levels,
            queries=[Q.equal("user_id", user_id), Q.limit(1)],
        )
        if res.get("documents"):
            doc = res["documents"][0]
            current_level = doc.get("current_level", 1)
            if level > current_level:
                db.update_document(DB_ID, settings.collection_user_levels, doc["$id"], {
                    "current_level": level,
                    "current_pulse": total_pulse,
                    "updated_at": now_iso,
                })
        else:
            db.create_document(DB_ID, settings.collection_user_levels, ID.unique(), {
                "user_id": user_id,
                "current_level": level,
                "current_pulse": total_pulse,
                "pulse_for_next": level * 100,
                "total_pulse_ever": total_pulse,
                "level_ups_count": 0,
                "prestige_rank": _get_prestige(total_pulse),
                "created_at": now_iso,
                "updated_at": now_iso,
            })
    except Exception as e:
        print(f"[!] Level check warning: {e}")


def _pulse_to_level(pulse: float) -> int:
    return max(1, min(150, int(pulse // 100) + 1))


def _get_prestige(pulse: float) -> str:
    if pulse >= 900:  return "Legend"
    if pulse >= 700:  return "Elite"
    if pulse >= 500:  return "Pro"
    if pulse >= 300:  return "Rising Star"
    if pulse >= 150:  return "Amateur"
    return "Rookie"


def _default_pulse(user_id: str, pulse: float = 100.0) -> dict:
    return {
        "user_id": user_id,
        "userId": user_id,
        "total_pulse": pulse,
        "totalPulse": pulse,
        "match_performance": 0,
        "matchPerformance": 0,
        "consistency": 0,
        "team_chemistry": 0,
        "teamChemistry": 0,
        "reliability": 0,
        "activity": 0,
        "leadership": 0,
        "tier": "contender",
        "level": 1,
        "level_progress_percent": 0,
        "prestige_rank": _get_prestige(pulse),
    }

File: src/services/pulseService.ts
/**
 * pulseService.ts
 * ─────────────────────────────────────────────────────────────────────────────
 * Frontend service for querying, updating, and syncing Appwrite `pulse_scores`
 * collection and `profiles.pulse_score` with live Realtime subscription.
 */

import { databases, client, ID, DATABASE_ID, COLLECTIONS, Query } from '@/lib/appwrite';
import type { PulseScore } from '@/types/pulse.types';

export interface PulseScoreRecord {
  $id?: string;
  user_id: string;
  total_pulse: number;
  match_performance: number;
  consistency: number;
  team_chemistry: number;
  reliability: number;
  activity: number;
  leadership: number;
  tier: 'contender' | 'elite' | 'pulse_elite';
  created_at?: string;
  updated_at?: string;
}

function calculateTier(score: number): 'CONTENDER' | 'ELITE' | 'PULSE ELITE' {
  if (score >= 900) return 'PULSE ELITE';
  if (score >= 800) return 'ELITE';
  return 'CONTENDER';
}

function toAppwriteTier(tier: 'CONTENDER' | 'ELITE' | 'PULSE ELITE'): 'contender' | 'elite' | 'pulse_elite' {
  switch (tier) {
    case 'PULSE ELITE': return 'pulse_elite';
    case 'ELITE': return 'elite';
    default: return 'contender';
  }
}

function fromAppwriteTier(tier: string | undefined): 'CONTENDER' | 'ELITE' | 'PULSE ELITE' {
  if (!tier) return 'CONTENDER';
  const clean = tier.toLowerCase().replace(/[^a-z]/g, '');
  if (clean === 'pulseelite' || clean === 'pulse_elite') return 'PULSE ELITE';
  if (clean === 'elite') return 'ELITE';
  return 'CONTENDER';
}

/**
 * Fetch the user's pulse_scores record from Appwrite.
 * If record does not exist yet, initializes it from profile or default.
 */
export async function getPulseScore(userId: string): Promise<PulseScore> {
  if (!userId) {
    return {
      userId: '',
      score: 100,
      tier: 'CONTENDER',
      breakdown: { matchPerf: 0, consistency: 0, chemistry: 0, reliability: 0, activity: 0, leadership: 0 },
      history: [],
      lastUpdated: new Date().toISOString(),
    };
  }

  try {
    const res = await databases.listDocuments(
      DATABASE_ID,
      COLLECTIONS.PULSE_SCORES,
      [Query.equal('user_id', userId), Query.limit(1)]
    );

    if (res.documents && res.documents.length > 0) {
      const doc = res.documents[0] as any;
      const score = Math.round(Number(doc.total_pulse ?? 100));
      const tier = fromAppwriteTier(doc.tier);

      // Fetch history if available
      let historyItems: any[] = [];
      try {
        const histRes = await databases.listDocuments(
          DATABASE_ID,
          'pulse_history',
          [Query.equal('user_id', userId), Query.orderDesc('created_at'), Query.limit(10)]
        );
        historyItems = histRes.documents.map((h: any) => ({
          date: (h.created_at || h.$createdAt || '').split('T')[0],
          score: Math.round(Number(h.score_after ?? score)),
          delta: Number(h.delta ?? 0),
          matchId: h.reference_id || h.$id,
        }));
      } catch {}

      return {
        userId,
        score,
        tier,
        breakdown: {
          matchPerf: Math.round(Number(doc.match_performance ?? 0)),
          consistency: Math.round(Number(doc.consistency ?? 0)),
          chemistry: Math.round(Number(doc.team_chemistry ?? 0)),
          reliability: Math.round(Number(doc.reliability ?? 0)),
          activity: Math.round(Number(doc.activity ?? 0)),
          leadership: Math.round(Number(doc.leadership ?? 0)),
        },
        history: historyItems.length > 0 ? historyItems : [
          { date: new Date().toISOString().split('T')[0], score, delta: 0, matchId: 'init' }
        ],
        lastUpdated: doc.updated_at || doc.$updatedAt || new Date().toISOString(),
      };
    }
  } catch (err) {
    console.warn('[pulseService] Error fetching pulse_scores from Appwrite:', err);
  }

  // Fallback: Check profile pulse_score and create initial pulse_scores document
  let profilePulse = 100;
  try {
    const profileDoc = await databases.getDocument(DATABASE_ID, COLLECTIONS.PROFILES, userId);
    if (profileDoc && profileDoc.pulse_score != null) {
      profilePulse = Number(profileDoc.pulse_score);
    }
  } catch {}

  // Create initial document in pulse_scores collection
  try {
    const nowIso = new Date().toISOString();
    await databases.createDocument(
      DATABASE_ID,
      COLLECTIONS.PULSE_SCORES,
      ID.unique(),
      {
        user_id: userId,
        total_pulse: profilePulse,
        match_performance: 0,
        consistency: 0,
        team_chemistry: 0,
        reliability: 0,
        activity: 0,
        leadership: 0,
        tier: toAppwriteTier(calculateTier(profilePulse)),
        created_at: nowIso,
        updated_at: nowIso,
      }
    );
  } catch (createErr) {
    console.warn('[pulseService] Error initializing pulse_scores record:', createErr);
  }

  return {
    userId,
    score: profilePulse,
    tier: calculateTier(profilePulse),
    breakdown: { matchPerf: 0, consistency: 0, chemistry: 0, reliability: 0, activity: 0, leadership: 0 },
    history: [{ date: new Date().toISOString().split('T')[0], score: profilePulse, delta: 0, matchId: 'init' }],
    lastUpdated: new Date().toISOString(),
  };
}

/**
 * Update pulse score and sub-scores in Appwrite pulse_scores collection,
 * synchronize profiles.pulse_score, and log to pulse_history.
 */
export async function updatePulseScore(
  userId: string,
  deltas: {
    matchPerf?: number;
    consistency?: number;
    chemistry?: number;
    reliability?: number;
    activity?: number;
    leadership?: number;
  },
  totalDelta: number,
  reason: string = 'Pulse Activity'
): Promise<PulseScore> {
  if (!userId) {
    throw new Error('User ID is required to update pulse score.');
  }

  const nowIso = new Date().toISOString();
  let existingDocId: string | null = null;
  let currentTotal = 100;
  let currentBreakdown = {
    matchPerf: 0,
    consistency: 0,
    chemistry: 0,
    reliability: 0,
    activity: 0,
    leadership: 0,
  };

  try {
    const res = await databases.listDocuments(
      DATABASE_ID,
      COLLECTIONS.PULSE_SCORES,
      [Query.equal('user_id', userId), Query.limit(1)]
    );
    if (res.documents && res.documents.length > 0) {
      const doc = res.documents[0] as any;
      existingDocId = doc.$id;
      currentTotal = Number(doc.total_pulse ?? 100);
      currentBreakdown = {
        matchPerf: Number(doc.match_performance ?? 0),
        consistency: Number(doc.consistency ?? 0),
        chemistry: Number(doc.team_chemistry ?? 0),
        reliability: Number(doc.reliability ?? 0),
        activity: Number(doc.activity ?? 0),
        leadership: Number(doc.leadership ?? 0),
      };
    }
  } catch {}

  const newTotal = Math.max(0, Math.min(1000, Math.round(currentTotal + totalDelta)));
  const newBreakdown = {
    matchPerf: Math.max(0, Math.min(100, Math.round(currentBreakdown.matchPerf + (deltas.matchPerf || 0)))),
    consistency: Math.max(0, Math.min(100, Math.round(currentBreakdown.consistency + (deltas.consistency || 0)))),
    chemistry: Math.max(0, Math.min(100, Math.round(currentBreakdown.chemistry + (deltas.chemistry || 0)))),
    reliability: Math.max(0, Math.min(100, Math.round(currentBreakdown.reliability + (deltas.reliability || 0)))),
    activity: Math.max(0, Math.min(100, Math.round(currentBreakdown.activity + (deltas.activity || 0)))),
    leadership: Math.max(0, Math.min(100, Math.round(currentBreakdown.leadership + (deltas.leadership || 0)))),
  };
  const newTier = calculateTier(newTotal);

  const payload = {
    user_id: userId,
    total_pulse: newTotal,
    match_performance: newBreakdown.matchPerf,
    consistency: newBreakdown.consistency,
    team_chemistry: newBreakdown.chemistry,
    reliability: newBreakdown.reliability,
    activity: newBreakdown.activity,
    leadership: newBreakdown.leadership,
    tier: toAppwriteTier(newTier),
    updated_at: nowIso,
  };

  // 1. Update or create in pulse_scores
  if (existingDocId) {
    try {
      await databases.updateDocument(DATABASE_ID, COLLECTIONS.PULSE_SCORES, existingDocId, payload);
    } catch (err) {
      console.error('[pulseService] Failed to update pulse_scores doc:', err);
    }
  } else {
    try {
      await databases.createDocument(DATABASE_ID, COLLECTIONS.PULSE_SCORES, ID.unique(), {
        ...payload,
        created_at: nowIso,
      });
    } catch (err) {
      console.error('[pulseService] Failed to create pulse_scores doc:', err);
    }
  }

  // 2. Synchronize profiles.pulse_score
  try {
    await databases.updateDocument(DATABASE_ID, COLLECTIONS.PROFILES, userId, {
      pulse_score: newTotal,
      updated_at: nowIso,
    });
  } catch (err) {
    console.warn('[pulseService] Failed to sync profile pulse_score:', err);
  }

  // 3. Log to pulse_history
  try {
    await databases.createDocument(DATABASE_ID, 'pulse_history', ID.unique(), {
      user_id: userId,
      delta: totalDelta,
      source: 'pulse_update',
      reason,
      score_after: newTotal,
      reference_id: `upd_${Date.now()}`,
      created_at: nowIso,
      updated_at: nowIso,
    });
  } catch (histErr) {
    console.warn('[pulseService] Failed to create pulse_history record:', histErr);
  }

  return {
    userId,
    score: newTotal,
    tier: newTier,
    breakdown: newBreakdown,
    history: [],
    lastUpdated: nowIso,
  };
}

/**
 * Subscribe to realtime updates on pulse_scores for immediate reflection when edited in Appwrite.
 */
export function subscribeToPulseScore(
  userId: string,
  onUpdate: (updatedScore: PulseScore) => void
): () => void {
  if (!userId) return () => {};

  const channel = `databases.${DATABASE_ID}.collections.${COLLECTIONS.PULSE_SCORES}.documents`;
  
  try {
    const unsubscribe = client.subscribe(channel, (response: any) => {
      const payload = response.payload;
      if (payload && payload.user_id === userId) {
        const score = Math.round(Number(payload.total_pulse ?? 100));
        const tier = fromAppwriteTier(payload.tier);
        onUpdate({
          userId,
          score,
          tier,
          breakdown: {
            matchPerf: Math.round(Number(payload.match_performance ?? 0)),
            consistency: Math.round(Number(payload.consistency ?? 0)),
            chemistry: Math.round(Number(payload.team_chemistry ?? 0)),
            reliability: Math.round(Number(payload.reliability ?? 0)),
            activity: Math.round(Number(payload.activity ?? 0)),
            leadership: Math.round(Number(payload.leadership ?? 0)),
          },
          history: [],
          lastUpdated: payload.updated_at || payload.$updatedAt || new Date().toISOString(),
        });
      }
    });

    return unsubscribe;
  } catch (err) {
    console.warn('[pulseService] Realtime subscription failed:', err);
    return () => {};
  }
}
